src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,148 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
+@SuppressWarnings("unchecked")
 public class ArrayList<T> implements List<T> {
+    private static int CAPACITY = 10;
+    private static final double CAPACITY_EXPAND_FACTOR = 1.5;
+    private Object[] objects;
+    private int size;
+
+    public ArrayList() {
+        objects = new Object[CAPACITY];
+    }
+
+    public ArrayList(List<T> list) {
+        objects = new Object[CAPACITY];
+        if (list.size() <= CAPACITY) {
+            copyListToArrayListAndResize(list);
+            return;
+        }
+        do {
+            expandObjectsArray(objects);
+        } while (list.size() <= CAPACITY);
+        copyListToArrayListAndResize(list);
+    }
+
     @Override
     public void add(T value) {
-
+        if (size < CAPACITY) {
+            objects[size] = value;
+            size++;
+            return;
+        }
+        expandObjectsArray(objects);
+        objects[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndex(index, 1);
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (size + 1 <= CAPACITY) {
+            copyArrayToObjectsAndResizeArrayList(
+                    objects, index, index + 1, size - index, 1);
+            objects[index] = value;
+            return;
+        }
+        expandObjectsArray(objects);
+        copyArrayToObjectsAndResizeArrayList(
+                objects, index, index + 1, size - index, 1);
+        objects[index] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        ArrayList<T> arrList = new ArrayList<>(list);
+        if (size + arrList.size <= CAPACITY) {
+            copyArrayToObjectsAndResizeArrayList(
+                    arrList.objects, 0, size, arrList.size, arrList.size);
+            return;
+        }
+        do {
+            expandObjectsArray(objects);
+        } while (size + arrList.size <= CAPACITY);
+        copyArrayToObjectsAndResizeArrayList(
+                arrList.objects, 0, size, arrList.size, arrList.size);
```

Comment:
use for loop to go through the passed list and add its elements using add method inside of this class

Reply:
The best!